### PR TITLE
Stop trying to estimate IL sizes

### LINKED ISSUE #36374: ASP.NET Core Dependency Injection container uses excessive amount of memory and CPU when large dependency graph is created
### Description
ASP.NET Core Dependency Injection container uses excessive amount of memory and CPU when large dependency graph is created.

### How To Reproduce
I created a [repository](https://github.com/sergey-kolodiy/AspNetCore.DI) with a small project that reproduces the issue. It targets .NET Framework 4.7.2, but the issue is also reproducible for .NET Core 2.2.

1. Build the solution and run `AspNetCore.DI.Default` project.
2. Using Fiddler or some other tool of your choice, send multiple simultaneous HTTP requests to `http://localhost:5000/api/values` API endpoint. The issue is not always reproduced if you send a single HTTP request and then just wait for the response.
3. Observe resources consumption. You should see significant increase in both CPU and memory consumption:
image
4. For comparison, run `AspNetCore.DI.Autofac` project that uses Autofac DI container instead of the default one and send multiple HTTP API requests again. You won't see those CPU and memory peaks this time.

### Expected behavior
Obviously the default DI container shouldn't use that much CPU and memory resources.

### So, what's going on?
Web API from the example above uses MVC middleware only, but, in addition to MVC services, it also registers a bunch of services with `Scoped` lifetime that are organized in a large graph (without any circular dependencies), and it seems like the default ASP.NET Core DI container is having some issues handling it.

In the request processing thread, the dependency graph is created and injected into the `ValuesController` just fine, and the API returns HTTP 200. But, it looks like for every dependency graph encountered, default DI container (or ASP.NET Core runtime) also spawns a separate background thread that performs some sort of initial processing and/or caching of that dependency graph. So, in theory this operation shouldn't affect request processing threads (at least when the dependency graph is relatively small).

However, it seems like this graph processing algorithm complexity grows exponentially, so the bigger dependency graph is, the more time and memory is consumed by this background operation. On my machine I had peaks up to 100% CPU (completely freezed) and more that 11GB memory consumption, which is not acceptable for production.

After this background process is done, resource consumption slowly decreases back to normal, and it doesn't start again when I invoke API methods on that controller, but for other controllers with different dependency graphs it starts again.

This issue is not reproducible when using Autofac DI container.

### PR DESCRIPTION
**Pull request by davidfowl**
- We originally did this to reduce the amount of resizes when generating IL for huge method bodies but now that we don't inline the transitive closure of scoped dependencies, we can just estimate a modest size that avoids *most* resizes (512).
- The current ILEmitCallsiteAnalyzer doesn't keep track of visited callsites so can over estimate the size. Instread of fixing it, just delete it.

Fixes https://github.com/dotnet/runtime/issues/36374

cc @brandondahler

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs [hunk 1]
```diff
@@ -100,8 +100,10 @@ private GeneratedMethod BuildTypeNoCache(ServiceCallSite callSite)
                 owner: GetType(),
                 skipVisibility: true);
 
-            ILEmitCallSiteAnalysisResult info = ILEmitCallSiteAnalyzer.Instance.CollectGenerationInfo(callSite);
-            ILGenerator ilGenerator = dynamicMethod.GetILGenerator(info.Size);
+            // In traces we've seen methods range from 100B - 4K sized methods since we've
+            // stop trying to inline everything into scoped methods. We'll pay for a couple of resizes
+            // so there'll be allocations but we could potentially change ILGenerator to use the array pool
+            ILGenerator ilGenerator = dynamicMethod.GetILGenerator(512);
```
Comment (benjiro) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs:106]: Is it worth moving this magic number out into a const and adding a further remark of why 512 was chosen?

**Review by benjiro — APPROVED** [review]
LGTM. Thanks for resolving this issue so quickly 🏆

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs (hunk 1 above)
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs:106]: I would normally agree if it used in more than on place. I could add size: 512 to make it a bit clearer but the comment is inline.

**Review by eerhardt — APPROVED** [review]

**Comment by davidfowl** [comment]
@eerhardt remind me what the process for backporting is? We have this bot now right

File: src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs [hunk 2]
```diff
@@ -100,8 +100,10 @@ private GeneratedMethod BuildTypeNoCache(ServiceCallSite callSite)
                 owner: GetType(),
                 skipVisibility: true);
 
-            ILEmitCallSiteAnalysisResult info = ILEmitCallSiteAnalyzer.Instance.CollectGenerationInfo(callSite);
-            ILGenerator ilGenerator = dynamicMethod.GetILGenerator(info.Size);
+            // In traces we've seen methods range from 100B - 4K sized methods since we've
+            // stop trying to inline everything into scoped methods. We'll pay for a couple of resizes
+            // so there'll be allocations but we could potentially change ILGenerator to use the array pool
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs:105]: > but we could potentially change ILGenerator to use the array pool

I'm skeptical we would do this.
https://github.com/dotnet/designs/blob/main/accepted/2021/runtime-security-mitigations.md#protect-intermediate-language-il-code
It introduces potential issues where an array is erronously returned to the pool while still in use, then the ILGenerator picks up that array and uses it to write IL, and now whoever returned the array to the pool is potentially able to write IL that'll be executed.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ILEmit/ILEmitResolverBuilder.cs:105]: Sure, I don't think it matters anymore really. We don't generate large methods anymore anyways 😄. Also now that we have this event, we can ask people to turn it on and see what sizes we get in practice to adjust this number.

**Pull request merged**